swarm/src/stream.rs

```diff
@@ -1,16 +1,40 @@
 use futures::{AsyncRead, AsyncWrite};
 use libp2p_core::muxing::SubstreamBox;
 use libp2p_core::Negotiated;
-use std::io::{IoSlice, IoSliceMut};
-use std::pin::Pin;
-use std::task::{Context, Poll};
+use std::{
+    io::{IoSlice, IoSliceMut},
+    pin::Pin,
+    sync::{Arc, Weak},
+    task::{Context, Poll},
+};
 
 #[derive(Debug)]
-pub struct Stream(Negotiated<SubstreamBox>);
+pub struct Stream {
+    stream: Negotiated<SubstreamBox>,
+    stream_counter: StreamCounter,
+}
+
+#[derive(Debug)]
+pub enum StreamCounter {
+    Arc(Arc<()>),
+    Weak(Weak<()>),
+}
 
 impl Stream {
-    pub(crate) fn new(stream: Negotiated<SubstreamBox>) -> Self {
-        Self(stream)
+    pub(crate) fn new(stream: Negotiated<SubstreamBox>, stream_counter: StreamCounter) -> Self {
```

Comment:
Streams are always initialized with a strong reference right, so we should change this parameter such that any outside module doesn't need to know about `StreamCounter`.

```suggestion
    pub(crate) fn new(stream: Negotiated<SubstreamBox>, counter: Arc<()>) -> Self {
```

Reply:
We can then make `StreamCounter` private to this module.

Reply:
fixed [8f09bb3](https://github.com/libp2p/rust-libp2p/pull/4595/commits/8f09bb347b9c971b25e2fdfcfab21bc4ef4a41f2)